### ISSUE TITLE
Give Interop the Ability to Reason About Types

### ISSUE THREAD

**Issue by iamrecursion** [oracle/graal#3405]
**Feature request**

Please include the following information:

**Is your feature request related to a problem? Please describe.**
Some languages (e.g. Python) have semantics that require verifying that an exception type actually _is_ an exception statically. Currently this means that the only way that GraalPython can catch polyglot exceptions is via a catch-all block, which is far less than ideal!

At the same time, other languages (such as [Enso](https://github.com/enso-org/enso)) have very different exception handling mechanisms where it would be useful to know that a given type can be used as an exception. It would be fantastic to be able to reason statically about the properties of _types_ and not just _values_ while using interop.

**Describe the solution you'd like.**
Extend the `InteropLibrary` protocol with a set of messages for reasoning about _types_ and not just _values_. While my primary motivating use case is something like `boolean InteropLibrary::isExceptionType(Object reciever)`, the same idea could be extended to the other types that `InteropLibrary` can talk about (`isNumberType`, `isArrayType`, and so on).

**Describe who do you think will benefit the most.**
The benefits of this would very much be trickle-down. While the primary beneficiary would be language implementors, as this would allow them to provide even more seamless interoperability between languages, these benefits trickle down to users of the integrated systems. The tighter the integration (without reaching the point of coupling), the more implementation details can be hidden, and the less the user has to know about what is going on under the hood!

**Describe alternatives you've considered.**
Unfortunately I can't think of another way of doing it that wouldn't break the encapsulation provided by the current interop abstraction. Maybe I'm not creative enough! ^^

**Additional context.**
This feature request was spurred by a discussion with @timfel on the slack. The relevant parts are quoted below for posterity due to the slack message limit.

> @timfel: the latter. python semantics require to raise a TypeError if a type named in an except clause is not an exception type
> 
> @timfel: but in interop, i can only check if instances are exceptions
>
> @iamrecursion: Are there any plans to extend interop to reason about types and not just values?
> 
> @iamrecursion: I know that we'd have a use for it as well. 
>
> @timfel: not at the moment, no

**Express whether you'd like to help contributing this feature**
I don't think I'd have time in the near term, unfortunately.

**Comment by chumer** [comment]
Thank you for the suggestion! I was thinking about this for a while, but was missing driving use-cases for this so far.

Interop has some existing ways to reason about types or how interop calls them meta-objects. 

To request the type of a value we have:
`InteropLibrary.hasMetaObject`
`InteropLibrary.getMetaObject`

The returned meta-object returns true for `InteropLibrary.isMetaObject`.

A meta-object provides currently provides the following capabilities:
getMetaSimpleName
getMetaQualifiedName 
isMetaInstance - to check whether an instance if of a type

What is missing is to understand whether two types/meta-objects are an instance of each other and whether meta-objects are specific known types.

To check whether meta-objects represent a specific known type it would be necessary to duplicate all interop types and trait messages for meta-objects. In order to avoid this I'd like to introduce prototypes instead of types. Every meta object can provide a prototype object that will then respond to the standard interop type and trait messages.

For example:

```
Object pythonValue;
InteropLibrary interop;

Object pythonMetaObject = interop.getMetaObject(pythonValue);
Object metaPrototype = interop.getMetaPrototype(pythonMetaObject);
interop.isNumber(metaPrototype) // is number
interop.isException(metaPrototype) // is exception
interop.hasArrayElements(metaPrototype) // is array
```

This has another nice side-effect that we can use the same prototype object for e.g. array component types:

```
Object arrayComponent = interop.getArrayComponent(array);
interop.fitsInInt(arrayComponent); // all components fit reliably in int! hooray!
```

Note that the meta prototype cannot be the meta-object itself because that would be incompatible (and probably quite confusing).

For interop proposal we need motivating examples for at least two, better three languages. Please post concrete examples here if you encounter them.

**Comment by eregon** [comment]
Wouldn't it be a problem for interop contracts?
For example if isException() is true it's documented that throwException must be supported, but it wouldn't be here.
Similar for `isNumber` where no asInt/asDouble/etc would work.

**Comment by iamrecursion** [comment]
This would definitely work for the kind of features I was envisaging, but I'm very curious about what it would mean for the interop contracts as they currently stand?

**Comment by chumer** [comment]
> Wouldn't it be a problem for interop contracts?

Yes it is. I was thinking we could disallow such values as regular values. So we could soften regular contracts for such values and you are not supposed to call any other messages on it other than type/trait messages. But that is brittle and problematic to assert in the contracts (maybe with a wrapper when assertions are on?).

The alternative to use dedicated messages is still on the table.  I need to do some more research what other people did for this. Maybe there is a nice solution I am not seeing.

We also still need better descriptions for use-cases. So let us not get into too many details of the API yet.

**Comment by eregon** [comment]
For the use-case of catching foreign exceptions, wouldn't it be acceptable to simply omit the check for whether the given "catch class expression" is an exception class if it's foreign? One could still throw some error if the value of the expression is not a class at all, i.e., not a `isMetaObject()`. And then use `isMetaInstance()` to check if the exception is of that class/type and whether the exception should be caught by that catch clause.
cc @timfel 

FWIW the semantics in Ruby are that the "catch class expression" must be a `Module` and the test is done with `===`. That should probably work naturally if we implement `Module === foreign` as `isMetaObject()`, and `===` on such objects as `isMetaInstance()`, but we could also handle foreign objects explicitly there if needed.

**Comment by iamrecursion** [comment]
For Enso, we have some niceties around how we display certain types in the IDE, so our use-case for this would be a bit different. Currently (beyond a core set of primitives such as booleans and numbers) we have a generic catch all "type" and hence "colour" for how we display polyglot values in the IDE. If we're given the ability to inspect the value's meta prototype then we can do a much better job of this!

While we can, of course, do this from a purely _value_ standpoint, this becomes more difficult when considering values that fall into potentially multiple archetypes and I feel like meta prototypes would allow us to disambiguate these cases.
